WordPress/src/main/java/org/wordpress/android/models/networkresource/ListNetworkResource.kt

```diff
@@ -0,0 +1,55 @@
+package org.wordpress.android.models.networkresource
+
+import org.wordpress.android.util.AppLog
+
+sealed class ListNetworkResource<T : Any>(val data: List<T>) {
+    class Init<T : Any> : ListNetworkResource<T>(ArrayList()) {
+        override fun updatedListNetworkResource(map: (old: T) -> T?) = this
+    }
+
+    class Ready<T : Any>(data: List<T>) : ListNetworkResource<T>(data) {
+        override fun updatedListNetworkResource(map: (old: T) -> T?) = Ready(mapNotNull(map))
+    }
+
+    class Success<T : Any>(data: List<T>, val canLoadMore: Boolean = false) : ListNetworkResource<T>(data) {
+        override fun updatedListNetworkResource(map: (old: T) -> T?) = Success(mapNotNull(map), canLoadMore)
+    }
+
+    class Loading<T : Any> private constructor(data: List<T>, val loadingMore: Boolean) : ListNetworkResource<T>(data) {
+        constructor(previous: ListNetworkResource<T>, loadingMore: Boolean = false): this(previous.data, loadingMore)
+
+        override fun updatedListNetworkResource(map: (old: T) -> T?) = Loading(mapNotNull(map), loadingMore)
+    }
+
+    class Error<T : Any> private constructor(data: List<T>, val errorMessage: String?, val wasLoadingMore: Boolean)
+        : ListNetworkResource<T>(data) {
+        constructor(previous: ListNetworkResource<T>, errorMessage: String?, wasLoadingMore: Boolean = false)
+                : this(previous.data, errorMessage, wasLoadingMore)
+
+        override fun updatedListNetworkResource(map: (old: T) -> T?) =
+                Error(mapNotNull(map), errorMessage, wasLoadingMore)
+    }
+
+    fun shouldFetch(loadMore: Boolean): Boolean {
+        return when (this) {
+            is Init -> { // Not ready yet
+                // TODO: Don't use T.MAIN for the log
+                AppLog.e(AppLog.T.MAIN, "ListNetworkResource should be ready before fetching")
+                false
+            }
+            is Loading -> false // Already fetching
+            is Success -> if (loadMore) canLoadMore else true // Trying to load more or refreshing
```

Comment:
Yes. `loadMore` being `false` means that we are trying to fetch the first page (most likely due to pull to refresh) and since there is no request going on right now we are able to fetch it. Couple notes:

1. I am still in the middle of between calling this method `shouldFetch` and `canFetch` neither has the full meaning of what I want.
2. There might be a better way to write this with Kotlin that'll be easier to understand at a first glance. I just didn't spend much time thinking about that as I was trying to see if it was working as quickly as possible.